// cosmo/lib/renderer.js
'use strict';

exports = module.exports;

var cmyk_to_hex = require( "colorvert/cmyk/hex" );

var shallowestOcean = {c:42,m:21,y:0,k:1};
var deepestOcean = {c:70,m:30,y:0,k:63};

var lowestValley = {c:0,m:11,y:32,k:71};
var tallestMountain = {c:0,m:10,y:30,k:4};

var weakNucium = {c:42,m:21,y:0,k:11};
var strongNucium = {c:70,m:35,y:0,k:43};

var weakNutro = {c:0,m:41,y:39,k:16};
var strongNutro = {c:0,m:61,y:58,k:34};

var sparseVegetation = {c:37,m:0,y:60,k:37};
var thickVegetation = {c:50,m:0,y:81,k:55};

exports.render = function(req,res)
{
	calculateHighestAndLowest(req.ctx);

	var colors = new Array(req.ctx.area);
	if(req.mode == "Depth")
	{
		for(var z = 0 ; z < req.ctx.area ; z++)
		{
			if(req.ctx.depth[z]>1)
			{
				colors[z] = cmykToHex(colorizeWater(req.ctx.depth[z],req.ctx));
			}
			else
			{
				colors[z] = cmykToHex(colorizeEarth(0,req.ctx));
			}
		}
	}
	else if (req.mode == "Height")
	{
		for(var z = 0 ; z < req.ctx.area ; z++)
		{
			//console.log(req.ctx.height[z]);
			colors[z] = cmykToHex(colorizeEarth(req.ctx.height[z],req.ctx));
		}
	}
	else if (req.mode == "Satellite")
	{
		for(var z = 0 ; z < req.ctx.area ; z++)
		{
			if(req.ctx.depth[z]>1)
			{
				colors[z] = cmykToHex(colorizeWater(req.ctx.depth[z],req.ctx));
			}
			else
			{
				colors[z] = cmykToHex(
					colorizeVegetationAndHeight(
						Math.floor(Math.random()*9),
						req.ctx.height[z],req.ctx));
			}
		}
	}
	else if (req.mode == "Elevation")
	{
		for(var z=0; z < req.ctx.area; z++)
		{
			colors[z] = cmykToHex(colorizeElevation(req.ctx.height[z] + req.ctx.depth[z],req.ctx));
		}
	}
	else if (req.mode == "Nutro")
	{
		for(var z=0; z < req.ctx.area; z++)
		{
			if(req.ctx.depth[z]>1)
			{
				colors[z] = cmykToHex(colorizeWater(req.ctx.depth[z],req.ctx));
			}
			else
			{
				colors[z] = cmykToHex(colorizeNutro(req.ctx.soilNutro[z],req.ctx));
			}
		}
	}
	else if (req.mode == "Nucium")
	{
		for(var z=0; z < req.ctx.area; z++)
		{
			if(req.ctx.depth[z]>1)
			{
				colors[z] = cmykToHex(colorizeWater(req.ctx.depth[z],req.ctx));
			}
			else
			{
				colors[z] =cmykToHex( colorizeNucium(req.ctx.soilNucium[z],req.ctx));
			}
		}
	}
	else if (req.mode == "Nutrients")
	{
		for(var z=0; z < req.ctx.area; z++)
		{
			if(req.ctx.depth[z]>1)
			{
				colors[z] = cmykToHex(colorizeWater(req.ctx.depth[z],req.ctx));
			}
			else
			{
				colors[z] = cmykToHex(colorizeNutrients(req.ctx.soilNutro[z], req.ctx.soilNucium[z],req.ctx));
			}
		}
	}

	else {
		res(req.mode + " doesn't exist.");
		return;
	}
	res(null, colors);
};
function colorValueBetween (value,minValue,maxValue,bottomColor,topColor) {
	var ratio = (value-minValue)/(maxValue-minValue);

	var deltaColor = {
		c:bottomColor.c-topColor.c,
		m:bottomColor.m-topColor.m,
		y:bottomColor.y-topColor.y,
		k:bottomColor.k-topColor.k
	};

	var cymk = {
		c: Math.round(bottomColor.c - (deltaColor.c*ratio)),
		m: Math.round(bottomColor.m - (deltaColor.m*ratio)),
		y: Math.round(bottomColor.y - (deltaColor.y*ratio)),
		k: Math.round(bottomColor.k - (deltaColor.k*ratio))
	};
	return cymk;
}
function cmykToHex(cmyk){
	
    var c = cmyk.c / 100;
    var m = cmyk.m / 100;
    var y = cmyk.y / 100;
    var k = cmyk.k / 100;
	
	var result = {};
    result.r = 1 - Math.min( 1, c * ( 1 - k ) + k );
    result.g = 1 - Math.min( 1, m * ( 1 - k ) + k );
    result.b = 1 - Math.min( 1, y * ( 1 - k ) + k );

    return rgbToHex(
    	Math.round( result.r * 255 ), 
    	Math.round( result.g * 255 ), 
    	Math.round( result.b * 255 ));
}

function componentToHex(c) {
    var hex = c.toString(16);
    return hex.length == 1 ? "0" + hex : hex;
}

function rgbToHex(r, g, b) {
    return "#" + componentToHex(r) + componentToHex(g) + componentToHex(b);
}


function calculateHighestAndLowest(ctx)
{
	ctx.highest = 1;
    ctx.hottest = 1;
    ctx.deepest = 1;
    ctx.brighest = 1;
    ctx.wettest = 1;
    ctx.tallest = 1;

    for(var z = 0 ; z < ctx.area ; z++)
	{
		if(ctx.height[z] > ctx.highest)
		{
			ctx.highest = ctx.height[z];
		}
		if(ctx.heat[z] > ctx.hottest)
		{
			ctx.hottest = ctx.heat[z];
		}
		if(ctx.depth[z] > ctx.deepest)
		{
			ctx.deepest = ctx.depth[z];
		}
		if(ctx.sunlight[z] > ctx.brightest)
		{
			ctx.brightest = ctx.sunlight[z];
		}
		if(ctx.rainfall[z] > ctx.wettest)
		{
			ctx.wettest = ctx.rainfall[z];
		}
		if(ctx.height[z] + ctx.depth[z] > ctx.tallest)
		{
			ctx.tallest = ctx.height[z] + ctx.depth[z];
		}
	}
}

function addColors(a,b)
{
	return {c: a.c + b.c,
			m: a.m + b.m,
			y: a.y + b.y,
			k: a.k + b.k};
}

function colorizeNutro(value,ctx)
{
	return colorValueBetween(value,0,100,weakNutro,strongNutro);	
}
function colorizeNucium(value,ctx)
{
	return colorValueBetween(value,0,100,weakNucium,strongNucium);	
}
function colorizeNutrients(nutro,nucium,ctx)
{
	return addColors(colorizeNutro(nutro,ctx), colorizeNucium(nucium,ctx));
}

function colorizeVegetation(value,ctx)
{
	return colorValueBetween(value,0,ctx.plantColumnsPer*ctx.plantRowsPer,sparseVegetation,thickVegetation);
}

function colorizeVegetationAndHeight(vegetation,height,ctx)
{
	return addColors(colorizeVegetation(vegetation,ctx), colorizeEarth(height,ctx));
}

function colorizeElevation(value,ctx)
{
	return colorValueBetween(value,0,ctx.tallest,lowestValley,tallestMountain);	
}
function colorizeEarth(value,ctx)
{
	return colorValueBetween(value,0,ctx.highest,lowestValley,tallestMountain);
}
function colorizeWater(value,ctx)
{
	return colorValueBetween(value,0,ctx.deepest,shallowestOcean,deepestOcean);
}
