// cosmo/lib/requestManager.js
'use strict';

var simulate = require('./simulate.js');
var manager = require('./simulationManager.js');
var utility = require('./utility.js');
var requests = new Array();

exports.queueSimulationRequest = function(request,res) {
    
    if(request.name && request.days)
    {
        if(request.days >= 1)
        {
            //  Do we have that simulation?
            manager.getSimulation(request.name,function(err,data)
                {
                    if(err)
                    {
                        res(err);
                    }
                    else
                    {
                        requests.push(request);
                        res(null,requests);
                    }
                });       
        }
        else
        {
            res(request.name + " cannot be simulated for " + request.days + ". Year values must be above 0.", requests);
        }
    }
    else 
    {
        res("Bad or missing request data.",requests);
    }
};

exports.deleteSimulationRequestsForWorld = function(request,res) {
    
    for(var i = requests.length-1; i >= 0; --i) {
        var item = requests[i];
        // console.log(item);
        // console.log(request);

        if(item.name == request.name) {
            requests.splice(i, 1);
        }
    }
    res(null,requests);
};

exports.getSimulationRequests = function(res) {
    res(null,requests);
};

exports.clearSimulationRequests = function(res) {
    requests = new Array();
    res(null,requests);
};

function save(simulation)
{
    var date = "" + simulation.year +","+simulation.month + "," + simulation.day;
    simulation.dates.push({date:date,ctx: utility.cloneObject(simulation.ctx)});
};

exports.processSimulationRequests = function(res)
{
    for(var r = 0 ; r < requests.length; r++) 
    {
        manager.getSimulation(requests[r].name,function(err,simulation)
        {
            if(err)
            {
                console.log(err);
            }
            else
            {
                //  Actually simulate this
                for(var d = 0 ; d < requests[r].days ; d++)
                {
                    simulation.day++;


                    if(simulation.day > 30)
                    {
                        simulation.day = 1;
                        simulation.month ++;
                        if(simulation.month > 12)
                        {
                            simulation.month = 1;
                            simulation.year++;

                            //  Yearly logic
                            //  Monthly logic
                            simulate.onYear(simulation.ctx,function(err,nothing)
                            {
                                if(err)
                                {
                                    res("Problem with crossing the year");
                                    return;
                                }
                            });
                        }

                        //  Monthly logic
                        simulate.onMonth(simulation.ctx,function(err,nothing)
                        {
                            if(err)
                            {
                                res("Problem with crossing the month");
                                return;
                            }
                        });
                    }

                    //  Weekly Logic
                    if(simulation.day % 10 == 0)
                    {

                    }

                    //  Daily logic
                    simulate.simulateDay(simulation.ctx,function(err,nothing)
                    {
                        if(err)
                        {
                            res("Problem wiht simulating a day");
                            return;
                        }
                        else
                        {
                            save(simulation);
                        }

                    });
                    
                }
            }
        });
    }
    requests = new Array();
    res(null,requests);
};
