// cosmo/lib/simulationManager.js
'use strict';

var simulate = require('./simulate.js');
var context = require('./context.js');
var util = require('./utility.js');

var simulations = new Array();

exports.createSimulation = function(req, res) 
{   
    var simulation = {
        name:req.name
        ,columns:req.dimensions.columns
        ,rows:req.dimensions.rows
        ,dates:[]
    };

    simulate.prepareSimulation(
        context(
            req.dimensions.columns, req.dimensions.rows,
            req.tilt, req.rotation,
            req.rules),
        function(err,ctx)
        {
            if(err) { 
                res(err);
                return;
            }

            simulation.dates.push(ctx);
            simulations.push(simulation);

            res(null,simulations);
        });
};

exports.deleteSimulation = function(req,res) {
    for(var i = simulations.length-1; i >= 0; --i) {
        var item = simulations[i];

        if(item.name == req) {
            simulations.splice(i, 1);
        }
    }
    res(null,simulations);
};
exports.clearSimulations = function(res) {
    simulations= new Array();
    res(null,simulations);
};

//  Expensive do not use
exports.getSimulations = function(res) {
    res(null,simulations);
};

exports.getSimulationDescriptions = function(res) {
    var descriptions = new Array();
    for(var i = 0 ; i < simulations.length ; i++)
    {
        descriptions.push(
        {
            name: simulations[i].name
            ,day: simulations[i].dates[simulations[i].dates.length-1].day
            ,month: simulations[i].dates[simulations[i].dates.length-1].month
            ,year: simulations[i].dates[simulations[i].dates.length-1].year
            ,rules: simulations[i].dates[simulations[i].dates.length-1].rules
        });
    }

    res(null,descriptions);
};

exports.getSimulation=function (req,res)
{
    for(var s = 0 ; s < simulations.length;s++)
    {
        //console.log(simulations[s].name + " equals " + req + "?");
        if(simulations[s].name == req)
        {
            res(null, simulations[s]);
            return;
        }
    }
    res("Cannot find simulation named " + req.name + ".");
};

exports.getSimulationContext = function (req,res)
{
    exports.getSimulation(req.name,function(err,simulation)
    {
        if(err)
        {
            res(err);
        }
        else
        {
            if(req.year == undefined)
            {
                if(simulation.dates.length>0)
                {
                    res(null,simulation.dates[simulation.dates.length-1]);
                }
                else
                {
                    res("Cannot get simulation context of empty dates.");
                }
            }
            else
            {
                for( var d = simulation.dates.length-1 ; d >= 0 ; d--)
                {
                    if(simulation.dates[d].year == req.year)
                    {
                        if(simulation.dates[d].month == req.month)
                        {
                            if(simulation.dates[d].day == req.day)
                            {
                                res(null,simulation.dates[d]);
                            }   
                        }       
                    }
                }
                res("Simulation named " + req.name + " does not have date " + req.year + "y-" + req.month + "m-" + req.day + "d.");
            }
        }
    });
};

exports.getSimulationTimeline = function(req,res) {
    
    exports.getSimulation(req,function(err,simulation) {
        
        if(err) 
        { 
            res(err);
        }
        else
        {
            var dates = new Array();
            for (var d = 0 ;d < simulation.dates.length;d++)
            {
                dates.push(
                    {
                        year:simulation.dates[d].year
                        ,month:simulation.dates[d].month
                        ,day:simulation.dates[d].day
                    })
            }
            res(null, dates);
        }
    });
};

exports.getSimulationDescription = function(req,res) {

    // console.log("Get simulation description");
    // console.log(req);

    exports.getSimulation(req,function(err,simulation) {
        
        if(err) { 
            res(err);
        }
        else {

            res(null, {
                name: simulation.name
                ,day: simulation.dates[simulation.dates.length-1].day
                ,month: simulation.dates[simulation.dates.length-1].month
                ,year: simulation.dates[simulation.dates.length-1].year
                ,rules: simulation.dates[simulation.dates.length-1].rules
            });
        }
          
    });
};
